### PR TITLE
Improved data preprocessing module

### PR DESCRIPTION
**Pull request by thanasibakis**
The aim of this PR is to get the data download/pre-process script into the package, add more pre-processing that was being done elsewhere, and clean the code up.

The easiest way to see what's new is to look where the new `linmod.data` module is used:

- `exploration/demo/README.md` describes how to run this part of the pipeline, which now no longer requires a path reference to the script
- `exploration/demo/*.py` have shorter, cleaner code chunks for `data = ...`, because that processing was moved to the package

The major changes to `linmod.data` are:
- Configuration parameters are global constants now
- Data pre-processing will filter collection dates to be no later than the specified forecast date
- The offset from the forecast date is included in the pre-processed data as `fd_offset`

Smaller changes are
- `linmod.eval` had a couple bugs that snuck in

### CODE REVIEW

**Review by swo — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -8,12 +8,67 @@ This repo hosts work on modeling how the composition of viral lineages, such as
...
 See the READMEs of individual folders in `exploration` for their specific workflows.
 
+### Architecture
+
+The model is provided with lightly-preprocessed data of variant sequences from humans in the USA, from [Nextstrain](https://docs.nextstrain.org/projects/ncov/en/latest/reference/remote_inputs.html) ([data dictionary](https://docs.nextstrain.org/projects/ncov/en/latest/reference/metadata-fields.html)).
+A CSV is provided, with columns `date`, `fd_offset`, `division`, `lineage`, `count`.
+Rows are uniquely identified by `(date, division, lineage)`.
+`date` and `fd_offset` can be computed from each other, given the forecast date.
+
+Note that `date` is the sample collection date. `fd` refers to the forecast date. Sequences are filtered to have a collection date no later than the forecast date.
```
Comment (afmagee42) [inline README.md:23]:
```suggestion
Note that `date` is the sample collection date. `fd` refers to the forecast date. `fd_offset` is `date - fd` measured in days. Sequences are filtered to have a collection date no later than the forecast date.
```

File: README.md [hunk 2]
```diff
@@ -8,12 +8,67 @@ This repo hosts work on modeling how the composition of viral lineages, such as
...
+
+| date       | fd_offset | division     | lineage | count |
+| ---------- | ---------- | ------------ | ------- | ----- |
+| 2024-05-07 | -12        | Arizona      | 24A     | 1     |
+| 2024-05-04 | -15        | Pennsylvania | 24A     | 2     |
+| ...        | ...        | ...          | ...     | ...   |
+
+The model must output samples of population-level lineage proportions.
+A CSV should be provided, with columns `fd_offset`, `division`, `lineage`, `sample_index`, and `phi`, for `fd_offset = -30, ..., 14`.
```
Comment (afmagee42) [inline README.md:32]:
```suggestion
A CSV should be provided, with columns `fd_offset`, `division`, `lineage`, `sample_index`, and `phi` (the population proportion), for `fd_offset = -30, ..., 14`.
```

File: exploration/demo/README.md [hunk 3]
```diff
@@ -2,10 +2,12 @@
...
-1. Download data with `data/load_metadata.py > exploration/demo/metadata.csv`
+1. Download data with `python3 -m linmod.data > metadata.csv`
 2. Fit models
     1. `cd exploration/demo`
     2. [If `exploration/demo/out` does not exist] `mkdir out`
     3. `./fit_baseline_model.py metadata.csv > out/samples-baseline.csv`
     4. `./fit_indep_div_model.py metadata.csv > out/samples-id.csv`
 3. Evaluate models with `./evaluate.py metadata.csv`
+
+Or, run `make`
```
Comment (afmagee42) [inline exploration/demo/README.md:13]:
```suggestion
Or, run `make` from within `exploration/demo`
```

**Review by afmagee42 — CHANGES_REQUESTED** [review]
I like having the data-wrangling stuff in a module, and I like having some documentation about the architecture.

I did a run test where I found one small bug. `pyarrow` is missing from `poetry.lock` and thus the visualization script errors out. We should be good to go with one `poetry add pyarrow`

I added some truly small comments, but I leave that to you to take or leave.

**Review by afmagee42 — APPROVED** [review]

**Pull request merged**